File: 백준/Gold/1238. 파티/파티.py
import heapq

n, m, x = map(int, input().split())
adj1 = [[] for _ in range(n+1)]
adj2 = [[] for _ in range(n+1)]
for _ in range(m):
	u,v,w = map(int, input().split())
	adj1[u].append([w,v])
	adj2[v].append([w,u])
inf = float('inf')
dist1 = [inf for _ in range(n + 1)]
dist2 = [inf for _ in range(n + 1)]
dist1[x] = 0
dist2[x] = 0

def f(adj, dist):
	q = []
	heapq.heappush(q, [0, x])
	while q:
		now_w, now_v = heapq.heappop(q)
		if dist[now_v] != now_w: continue
		for nxt_w, nxt_v in adj[now_v]:
			if dist[nxt_v] <= dist[now_v] + nxt_w: continue
			dist[nxt_v] = dist[now_v] + nxt_w
			heapq.heappush(q, [dist[nxt_v], nxt_v])

f(adj1, dist1)
f(adj2, dist2)

ans = 0
for i in range(1, n+1):
	if dist1[i]==inf or dist2[i]==inf: continue
	ans = max(ans, dist1[i]+dist2[i])
print(ans)

File: 백준/Gold/1238. 파티/파티.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
    static StringTokenizer st;
    static int x;

    public static void main(String[] args) throws Exception {
        st = new StringTokenizer(input.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        x = Integer.parseInt(st.nextToken());

        ArrayList<int[]>[] adj1 = new ArrayList[n + 1];
        ArrayList<int[]>[] adj2 = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            adj1[i] = new ArrayList<>();
            adj2[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(input.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());

            adj1[a].add(new int[]{w, b});
            adj2[b].add(new int[]{w, a});
        }

        int[] dist1 = new int[n + 1];
        int[] dist2 = new int[n + 1];
        Arrays.fill(dist1, Integer.MAX_VALUE);
        Arrays.fill(dist2, Integer.MAX_VALUE);

        dijk(dist1, adj1);
        dijk(dist2, adj2);

        int ans = 0;
        for (int i = 1; i <= n; i++) {
            ans = Math.max(ans, dist1[i] + dist2[i]);
        }

        System.out.println(ans);
    }

    private static void dijk(int[] dist, ArrayList<int[]>[] adj) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[0] - o2[0]);
        pq.add(new int[]{0, x});
        dist[x] = 0;

        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int w = cur[0];
            int now = cur[1];

            if (w != dist[now]) {
                continue;
            }

            for (int[] next : adj[now]) {
                int nw = next[0];
                int nxt = next[1];
                if (dist[nxt] <= dist[now] + nw) {
                    continue;
                }

                dist[nxt] = dist[now] + nw;
                pq.add(new int[]{dist[nxt], nxt});
            }
        }


    }
}